src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -8938,74 +8894,77 @@ private ContinueStatementSyntax ParseContinueStatement(SyntaxList<AttributeListS
 
         private TryStatementSyntax ParseTryStatement(SyntaxList<AttributeListSyntax> attributes)
         {
-            var isInTry = _isInTry;
-            _isInTry = true;
+            Debug.Assert(this.CurrentToken.Kind is SyntaxKind.TryKeyword or SyntaxKind.CatchKeyword or SyntaxKind.FinallyKeyword);
 
+            // We are called into on try/catch/finally, so eating the try may actually fail.
             var @try = this.EatToken(SyntaxKind.TryKeyword);
 
-            BlockSyntax block;
+            BlockSyntax tryBlock;
             if (@try.IsMissing)
             {
-                block = _syntaxFactory.Block(
-                    attributeLists: default, this.EatToken(SyntaxKind.OpenBraceToken), default(SyntaxList<StatementSyntax>), this.EatToken(SyntaxKind.CloseBraceToken));
+                // If there was no actual `try`, then we got here because of a misplaced `catch`/`finally`.  In that
+                // case just synthesize a fully missing try-block.  We will already have issued a diagnostic on the
+                // `try` keyword, so we don't need to issue any more.
+
+                Debug.Assert(@try.ContainsDiagnostics);
+                Debug.Assert(this.CurrentToken.Kind is SyntaxKind.CatchKeyword or SyntaxKind.FinallyKeyword);
+
+                tryBlock = missingBlock();
             }
             else
             {
                 var saveTerm = _termState;
                 _termState |= TerminatorState.IsEndOfTryBlock;
-                block = this.ParsePossiblyAttributedBlock();
+                tryBlock = this.ParsePossiblyAttributedBlock();
                 _termState = saveTerm;
             }
 
-            var catches = default(SyntaxListBuilder<CatchClauseSyntax>);
-            FinallyClauseSyntax @finally = null;
+            SyntaxListBuilder<CatchClauseSyntax> catchClauses = default;
+            FinallyClauseSyntax finallyClause = null;
             try
             {
-                bool hasEnd = false;
-
                 if (this.CurrentToken.Kind == SyntaxKind.CatchKeyword)
                 {
-                    hasEnd = true;
-                    catches = _pool.Allocate<CatchClauseSyntax>();
+                    catchClauses = _pool.Allocate<CatchClauseSyntax>();
                     while (this.CurrentToken.Kind == SyntaxKind.CatchKeyword)
-                    {
-                        catches.Add(this.ParseCatchClause());
-                    }
+                        catchClauses.Add(this.ParseCatchClause());
```

Comment:
nit: this could just be how my brain works, I'm able to handle the `if` statements without braces, but `while` throws me off (no longer feels like a loop). Could we keep the braces here at least?

Reply:
sounds good.  added back :)